addons/stock/report/report_stock_reception.py

```diff
@@ -0,0 +1,283 @@
+            if qty_already_assigned:
+                product_to_total_assigned[move.product_id][0] += qty_already_assigned
+                product_to_total_assigned[move.product_id][1].append(move.id)
+            if move.product_qty != qty_already_assigned:
+                if move.state == 'draft':
+                    product_to_qty_draft[move.product_id] += move.product_qty - qty_already_assigned
+                else:
+                    quantity_to_assign = move.product_qty
+                    if move.picking_id.immediate_transfer:
+                        # if immediate transfer is not Done and quantity_done hasn't been edited, then move.product_qty will incorrectly = 1 (due to default)
+                        quantity_to_assign = move.product_uom._compute_quantity(move.quantity_done, move.product_id.uom_id, rounding_method='HALF-UP')
+                    product_to_qty_to_assign[move.product_id].append((quantity_to_assign - qty_already_assigned, move))
+
+        # only match for non-mto moves in same warehouse
+        warehouse = pickings[0].picking_type_id.warehouse_id
+        wh_location_ids = self.env['stock.location'].search([('id', 'child_of', warehouse.view_location_id.id), ('location_id.usage', '!=', 'supplier')]).ids
+
+        allowed_states = ['confirmed', 'partially_available', 'waiting']
+        if 'done' in picking_states:
+            # only done moves are allowed to be assigned to already reserved moves
+            allowed_states += ['assigned']
+
+        outs = self.env['stock.move'].search(
+            [
+                ('state', 'in', allowed_states),
+                ('product_qty', '>', 0),
+                ('location_id', 'in', wh_location_ids),
+                ('move_orig_ids', '=', False),
+                ('picking_id', 'not in', pickings.ids),
+                ('product_id', 'in',
+                    [p.id for p in list(product_to_qty_to_assign.keys()) + list(product_to_qty_draft.keys())]),
+            ],
+            order='reservation_date, priority desc, date, id')
+
+        products_to_outs = defaultdict(list)
+        for out in outs:
+            products_to_outs[out.product_id].append(out)
+
+        sources_to_lines = defaultdict(list)  # group by source so we can print them together
+        # show potential moves that can be assigned
+        for product_id, outs in products_to_outs.items():
+            for out in outs:
+                # only show pickings or MOs, no SOs
+                source = out.picking_id if out.picking_id else out._get_source_document()
+                if not source:
+                    continue
+
+                qty_to_reserve = out.product_qty
+                product_uom = out.product_id.uom_id
+                if 'done' not in picking_states and out.state == 'partially_available':
+                    qty_to_reserve -= out.product_uom._compute_quantity(out.reserved_availability, product_uom)
+                moves_in = self.env['stock.move']
```

Comment:
`moves_in = product_to_qty_to_assign[out.product_id][1]` (to avoid useless union of recordset `moves_in |= move_in`)